File: bundles/gitlab-eclipse-plugin/src/com/gitlab/eclipse/lsp/GitLabLanguageServerConfigurationParams.java
package com.gitlab.eclipse.lsp;

import java.util.List;

public record GitLabLanguageServerConfigurationParams
(
	String baseUrl,
	String logLevel,
	GitLabLanguageServerConfigurationParams.Telemetry telemetry,
	String token,
	GitLabLanguageServerConfigurationParams.CodeCompletion codeCompletion,
	GitLabLanguageServerConfigurationParams.FeatureFlags featureFlags,
	boolean ignoreCertificateErrors
	// String projectPath,
) {
	public static class Builder {
		private String baseUrl;
		private String logLevel;
		private GitLabLanguageServerConfigurationParams.CodeCompletion codeCompletion;
		private GitLabLanguageServerConfigurationParams.FeatureFlags featureFlags;
		private GitLabLanguageServerConfigurationParams.Telemetry telemetry;
		private String token;
		private boolean ignoreCertificateErrors;

		public Builder baseUrl(String baseUrl) {
			this.baseUrl = baseUrl;
			return this;
		}

		public Builder codeCompletion(GitLabLanguageServerConfigurationParams.CodeCompletion codeCompletion) {
			this.codeCompletion = codeCompletion;
			return this;
		}

		public Builder featureFlags(FeatureFlags featureFlags) {
			this.featureFlags = featureFlags;
			return this;
		}

		public Builder ignoreCertificateErrors(boolean ignoreCertificateErrors) {
			this.ignoreCertificateErrors = ignoreCertificateErrors;
			return this;
		}

		public Builder logLevel(String logLevel) {
			this.logLevel = logLevel;
			return this;
		}

		public Builder telemetry(GitLabLanguageServerConfigurationParams.Telemetry telemetry) {
			this.telemetry = telemetry;
			return this;
		}

		public Builder token(String token) {
			this.token = token;
			return this;
		}

		public GitLabLanguageServerConfigurationParams build() {
			return new GitLabLanguageServerConfigurationParams(baseUrl, logLevel, telemetry, token, codeCompletion, featureFlags, ignoreCertificateErrors);
		}
	}

	public static record CodeCompletion(boolean enableSecretRedaction, List<String> disabledSupportedLanguages, List<String> additionalLanguages) {}
	
	public static record FeatureFlags(boolean remoteSecurityScans, boolean streamCodeGenerations) {
		public static class Builder {
			boolean remoteSecurityScans;
			boolean streamCodeGenerations;

			public Builder remoteSecurityScans(boolean remoteSecurityScans) {
				this.remoteSecurityScans = remoteSecurityScans;
				return this;
			}

			public Builder streamCodeGenerations(boolean streamCodeGenerations) {
				this.streamCodeGenerations = streamCodeGenerations;
				return this;
			}

			public FeatureFlags build() {
				return new FeatureFlags(remoteSecurityScans, streamCodeGenerations);
			}
		}

		public static Builder builder() {
			return new Builder();
		}
	}

	public static record HttpAgentOptions(String ca, String cert, String certKey) {}

	public static record Telemetry(boolean enabled, String trackingUrl, List<TelemetryAction> actions) {}

	public static record TelemetryAction(String action) {}

	public static GitLabLanguageServerConfigurationParams.Builder builder() {
		return new Builder();
	}
}

File: src/main/kotlin/com/gitlab/eclipse/lsp/configuration/GitLabLanguageServerConfigurationParams.kt
package com.gitlab.eclipse.lsp.configuration

import com.gitlab.eclipse.telemetry.params.TelemetryAction
import org.eclipse.lsp4j.WorkspaceFolder

data class GitLabLanguageServerConfigurationParams(
  val baseUrl: String? = null,
  val logLevel: String? = null,
  val telemetry: Telemetry? = null,
  val token: String? = null,
  val codeCompletion: CodeCompletion? = null,
  val featureFlags: FeatureFlags? = null,
  val ignoreCertificateErrors: Boolean = false,
  val httpAgentOptions: HttpAgentOptions? = null,
  val workspaceFolders: List<WorkspaceFolder>? = null,
  val duoChat: DuoChat? = null,
  val duo: Duo? = null,
  val securityScannerOptions: SecurityScannerOptions? = null,
) {
  /**
   * 設計 §9.1。生成形は [token] を平文で載せる。成分は 12 個あるが生成形の再現は目指さず、
   * 診断に有用なものだけを選ぶ(設計 L-4)。[httpAgentOptions] は入れ子の `toString` に
   * 委譲する(設計 §9.1)— 中身に直接手を伸ばさない。
   */
  override fun toString(): String =
    "GitLabLanguageServerConfigurationParams(baseUrl=$baseUrl, logLevel=$logLevel, " +
      "token=${token?.let { "***" }}, ignoreCertificateErrors=$ignoreCertificateErrors, " +
      "httpAgentOptions=$httpAgentOptions)"

  data class CodeCompletion(
    val enabled: Boolean = true,
    val enableSecretRedaction: Boolean = true,
    val disabledSupportedLanguages: List<String> = emptyList(),
    val additionalLanguages: List<String> = emptyList()
  )

  data class FeatureFlags(val remoteSecurityScans: Boolean, val streamCodeGenerations: Boolean)

  // The language server reads only `securityScannerOptions.enabled` (design §6.1 P2 step 6).
  data class SecurityScannerOptions(val enabled: Boolean)

  data class DuoChat(val enabled: Boolean)

  // Spelling matters: the language server reads `enabledWithoutGitlabProject` (lowercase "l").
  data class Duo(
    val enabledWithoutGitlabProject: Boolean,
    // The language server's AgenticChatSupportCheck reads `duo.agentPlatform.enabled` and
    // defaults it to FALSE when absent (`get("duo.agentPlatform.enabled") ?? false`), skipping
    // the agentic availability check entirely. It must be sent explicitly for Agentic Chat to
    // ever become available. (`agent-platform-disabled-by-user` treats absent as enabled, so
    // the mismatch is invisible in feature-state checks.)
    val agentPlatform: AgentPlatform? = null,
  )

  data class AgentPlatform(val enabled: Boolean)

  data class HttpAgentOptions(val ca: String?, val cert: String? = null, val certKey: String? = null) {
    /**
     * 設計 §9.1。[ca] は名前パターンに当たらないが CA 証明書のパスまたは内容であり、
     * 設計 §7.2 の明示リストで秘匿と宣言している。
     */
    override fun toString(): String =
      "HttpAgentOptions(ca=${ca?.let { "***" }}, cert=${cert?.let { "***" }}, certKey=${certKey?.let { "***" }})"
  }

  data class Telemetry(
    val enabled: Boolean,
    val trackingUrl: String,

    // This structure is defined in https://gitlab.com/gitlab-org/editor-extensions/gitlab-lsp/-/blob/main/docs/supported_messages.md?ref_type=heads#code-suggestions-telemetry
    val actions: List<Map<String, String>> = TelemetryAction.entries.map { action -> mapOf("action" to action.value) }
  )
}
